File: src/services/x/timeline/xTimelineParser.ts
import {
  XFollowingTimelineBatch,
  XTimelineItem,
  XTimelineMedia,
  XTimelineMediaType,
} from './xTimelineTypes';

// Helpers below (isRecord, asRecord, asString, asNumber, toIsoDate) are new — Rettiwt-API
// uses TypeScript interfaces/casts on the raw JSON instead. These exist because this codebase
// treats the API response as `unknown` and validates at runtime rather than casting.

const isRecord = (value: unknown): value is Record<string, unknown> => {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value);
};

const asRecord = (value: unknown): Record<string, unknown> | null => {
  return isRecord(value) ? value : null;
};

const asString = (value: unknown): string => {
  return typeof value === 'string' ? value : '';
};

const asNumber = (value: unknown): number => {
  if (typeof value === 'number' && Number.isFinite(value)) {
    return value;
  }

  if (typeof value === 'string') {
    const parsed = Number(value);
    return Number.isFinite(parsed) ? parsed : 0;
  }

  return 0;
};

// Rettiwt-API does `new Date(tweet.legacy.created_at).toISOString()` inline in the Tweet
// constructor. This is the same logic extracted into a safe helper with NaN-checking.
const toIsoDate = (rawDate: unknown): string => {
  if (typeof rawDate !== 'string') {
    return '';
  }

  const parsed = new Date(rawDate);
  if (Number.isNaN(parsed.valueOf())) {
    return '';
  }

  return parsed.toISOString();
};

// Taken from Rettiwt-API's findByFilter (src/helper/JsonUtils.ts).
// Same recursive deep-search logic. Changed: accepts `unknown` instead of `NonNullable<unknown>`
// and uses isRecord guard instead of raw typeof checks for stricter runtime safety.
// Added: soft node/collection limits to prevent pathological deep scans on mobile.
const MAX_FILTER_MATCHES = 1000;
const MAX_FILTER_NODES = 15000;

const findByFilter = <T>(data: unknown, key: string, value: string): T[] => {
  const result: T[] = [];
  let visitedNodes = 0;

  const visit = (node: unknown) => {
    if (visitedNodes >= MAX_FILTER_NODES || result.length >= MAX_FILTER_MATCHES) {
      return;
    }

    visitedNodes += 1;

    if (Array.isArray(node)) {
      for (const item of node) {
        visit(item);
        if (visitedNodes >= MAX_FILTER_NODES || result.length >= MAX_FILTER_MATCHES) {
          return;
        }
      }
      return;
    }

    if (!isRecord(node)) {
      return;
    }

    if (Object.prototype.hasOwnProperty.call(node, key) && node[key] === value) {
      result.push(node as unknown as T);
      if (result.length >= MAX_FILTER_MATCHES) {
        return;
      }
    }

    for (const valueOfKey of Object.values(node)) {
      visit(valueOfKey);
      if (visitedNodes >= MAX_FILTER_NODES || result.length >= MAX_FILTER_MATCHES) {
        return;
      }
    }
  };

  visit(data);
  return result;
};

// Maps raw X API media type strings to XTimelineMediaType.
// Rettiwt-API uses RawMediaType enum -> MediaType enum mapping in TweetMedia constructor.
// Changed: keeps the raw string values directly and adds 'unknown' fallback.
const mediaTypeFromRaw = (value: unknown): XTimelineMediaType => {
  if (value === 'photo' || value === 'video' || value === 'animated_gif') {
    return value;
  }

  return 'unknown';
};

// Based on Rettiwt-API's TweetMedia constructor (src/models/data/Tweet.ts TweetMedia).
// Same approach: photos use media_url_https, videos pick the highest-bitrate mp4 variant.
// Changed: also checks content_type for 'mp4' (Rettiwt-API only checks bitrate).
// GIFs are not special-cased (Rettiwt-API picks variants[0] for GIFs); instead they go
// through the same best-bitrate loop, which works because GIFs typically have one variant.
const extractBestMediaUrl = (
  media: Record<string, unknown>,
): { url: string; thumbnailUrl?: string } => {
  const type = mediaTypeFromRaw(media.type);

  if (type === 'photo') {
    return { url: asString(media.media_url_https) };
  }

  const videoInfo = asRecord(media.video_info);
  const variants = Array.isArray(videoInfo?.variants) ? videoInfo.variants : [];
  let selectedUrl = '';
  let selectedBitrate = -1;

  for (const variant of variants) {
    const record = asRecord(variant);
    if (!record) {
      continue;
    }

    const contentType = asString(record.content_type);
    const url = asString(record.url);
    const bitrate = asNumber(record.bitrate);

    if (!url) {
      continue;
    }

    if (contentType.includes('mp4') && bitrate >= selectedBitrate) {
      selectedUrl = url;
      selectedBitrate = bitrate;
      continue;
    }

    if (!selectedUrl) {
      selectedUrl = url;
    }
  }

  return {
    url: selectedUrl,
    thumbnailUrl: asString(media.media_url_https) || undefined,
  };
};

// Based on Rettiwt-API's Tweet constructor media extraction:
//   `tweet.legacy.extended_entities?.media?.map(m => new TweetMedia(m))`
// Same source path (legacy.extended_entities.media). Changed: adds expandedUrl field,
// filters out entries with no url and no expandedUrl, and conditionally sets thumbnailUrl.
const extractMedia = (legacy: Record<string, unknown>): XTimelineMedia[] => {
  const extendedEntities = asRecord(legacy.extended_entities);
  const mediaEntries = Array.isArray(extendedEntities?.media) ? extendedEntities.media : [];

  return mediaEntries
    .map((entry) => {
      const media = asRecord(entry);
      if (!media) {
        return null;
      }

      const type = mediaTypeFromRaw(media.type);
      const { url, thumbnailUrl } = extractBestMediaUrl(media);
      const expandedUrl = asString(media.expanded_url);

      if (!url && !expandedUrl) {
        return null;
      }

      const item: XTimelineMedia = {
        type,
        url,
        expandedUrl,
      };

      if (thumbnailUrl) {
        item.thumbnailUrl = thumbnailUrl;
      }

      return item;
    })
    .filter((media): media is XTimelineMedia => media !== null);
};

// Based on Rettiwt-API's Tweet.timeline() + Tweet constructor (src/models/data/Tweet.ts).
// The TweetWithVisibilityResults handling mirrors Rettiwt-API's _getQuotedTweet/_getRetweetedTweet.
// Changed: returns a flat XTimelineItem instead of a Tweet class instance.
// Extracts isRetweet/isQuote as booleans instead of nesting full quoted/retweeted Tweet objects.
// Uses note_tweet for long-form text the same way as Rettiwt-API's Tweet constructor.
// URL construction matches Rettiwt-API: `https://x.com/${userName}/status/${id}`.
const extractTweet = (timelineTweetNode: unknown): XTimelineItem | null => {
  const timelineRecord = asRecord(timelineTweetNode);
  const tweetResults = asRecord(timelineRecord?.tweet_results);
  let result = asRecord(tweetResults?.result);

  if (!result) {
    return null;
  }

  // From Rettiwt-API: handles TweetWithVisibilityResults by unwrapping to .tweet
  if (result.__typename === 'TweetWithVisibilityResults') {
    result = asRecord(result.tweet);
  }

  if (!result) {
    return null;
  }

  const legacy = asRecord(result.legacy);
  const id = asString(result.rest_id);
  if (!legacy || !id) {
    return null;
  }

  // From Rettiwt-API Tweet constructor: prefers note_tweet text over legacy.full_text
  const noteTweet = asRecord(result.note_tweet);
  const noteTweetResults = asRecord(noteTweet?.note_tweet_results);
  const noteTweetResult = asRecord(noteTweetResults?.result);

  // From Rettiwt-API: user info extracted via core.user_results.result
  const core = asRecord(result.core);
  const userResults = asRecord(core?.user_results);
  const userResult = asRecord(userResults?.result);
  const userLegacy = asRecord(userResult?.legacy);
  const authorHandle = asString(userLegacy?.screen_name);

  // From Rettiwt-API _getQuotedTweet / _getRetweetedTweet — simplified to boolean flags
  const quotedStatusResult = asRecord(result.quoted_status_result);
  const quotedResult = asRecord(quotedStatusResult?.result);
  const retweetedStatusResult = asRecord(legacy.retweeted_status_result);
  const retweetedResult = asRecord(retweetedStatusResult?.result);
  const retweetedResultTweet = asRecord(retweetedResult?.tweet);

  const isRetweet = Boolean(
    asString(retweetedResult?.rest_id) || asString(retweetedResultTweet?.rest_id),
  );
  const isQuote = Boolean(legacy.is_quote_status || asString(quotedResult?.rest_id));

  let quotedText = '';
  let quotedAuthorHandle = '';
  if (isQuote && quotedResult) {
    const qLegacy = asRecord(quotedResult.legacy);
    const qNoteTweet = asRecord(quotedResult.note_tweet);
    const qNoteResults = asRecord(qNoteTweet?.note_tweet_results);
    const qNoteResult = asRecord(qNoteResults?.result);
    quotedText = asString(qNoteResult?.text) || asString(qLegacy?.full_text);

    const qCore = asRecord(quotedResult.core);
    const qUserResults = asRecord(qCore?.user_results);
    const qUserResult = asRecord(qUserResults?.result);
    const qUserLegacy = asRecord(qUserResult?.legacy);
    quotedAuthorHandle = asString(qUserLegacy?.screen_name);
  }

  return {
    id,
    text: asString(noteTweetResult?.text) || asString(legacy.full_text),
    createdAt: toIsoDate(legacy.created_at),
    authorName: asString(userLegacy?.name),
    authorHandle,
    lang: asString(legacy.lang),
    replyTo: asString(legacy.in_reply_to_status_id_str),
    quoteCount: asNumber(legacy.quote_count),
    replyCount: asNumber(legacy.reply_count),
    retweetCount: asNumber(legacy.retweet_count),
    likeCount: asNumber(legacy.favorite_count),
    viewCount: asString(asRecord(result.views)?.count)
      ? asNumber(asRecord(result.views)?.count)
      : null,
    isRetweet,
    isQuote,
    quotedText,
    quotedAuthorHandle,
    url: authorHandle ? `https://x.com/${authorHandle}/status/${id}` : '',
    media: extractMedia(legacy),
  };
};

// Based on Rettiwt-API's Tweet.timeline() + CursoredData constructor (src/models/data/CursoredData.ts).
// Tweet extraction uses findByFilter(__typename, 'TimelineTweet') — same as Rettiwt-API.
// Cursor extraction uses findByFilter(cursorType, 'Bottom') — same as Rettiwt-API.
// Changed: added deduplication by tweet id (not in Rettiwt-API). Returns null instead of
// empty string when no next cursor exists.
export const parseXFollowingTimelineResponse = (payload: unknown): XFollowingTimelineBatch => {
  const timelineTweetNodes = findByFilter<unknown>(payload, '__typename', 'TimelineTweet');
  const deduped = new Map<string, XTimelineItem>();

  for (const node of timelineTweetNodes) {
    const item = extractTweet(node);
    if (!item || !item.id || deduped.has(item.id)) {
      continue;
    }

    deduped.set(item.id, item);
  }

  const cursors = findByFilter<Record<string, unknown>>(payload, 'cursorType', 'Bottom');
  const nextCursor = asString(cursors[0]?.value) || null;

  return {
    items: Array.from(deduped.values()),
    nextCursor,
  };
};

File: android/app/src/main/kotlin/com/yonisirote/readmyfeed/providers/x/timeline/XTimelineParser.kt
package com.yonisirote.readmyfeed.providers.x.timeline

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.time.Instant
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter
import java.util.Locale

private const val MAX_FILTER_MATCHES = 1000
private const val MAX_FILTER_NODES = 15000

private val parserJson = Json { ignoreUnknownKeys = true }
private val xDateFormatter = DateTimeFormatter.ofPattern("EEE MMM dd HH:mm:ss Z yyyy", Locale.ENGLISH)

fun parseXFollowingTimelineResponse(payload: String): XFollowingTimelineBatch {
  val root = try {
    parserJson.parseToJsonElement(payload)
  } catch (error: Exception) {
    throw XTimelineException(
      message = "Timeline response was not valid JSON.",
      code = XTimelineErrorCodes.RESPONSE_INVALID,
      context = mapOf("cause" to (error.message ?: error::class.java.simpleName)),
      cause = error,
    )
  }

  return parseXFollowingTimelineResponse(root)
}

fun parseXFollowingTimelineResponse(payload: JsonElement): XFollowingTimelineBatch {
  val timelineTweetNodes = findByFilter(payload, "__typename", "TimelineTweet")
  val deduped = LinkedHashMap<String, XTimelineItem>()

  for (node in timelineTweetNodes) {
    val item = extractTweet(node) ?: continue
    if (item.id.isBlank() || deduped.containsKey(item.id)) {
      continue
    }
    deduped[item.id] = item
  }

  return XFollowingTimelineBatch(
    items = deduped.values.toList(),
    nextCursor = extractNextCursor(payload),
  )
}

private fun findByFilter(data: JsonElement, key: String, value: String): List<JsonObject> {
  val result = mutableListOf<JsonObject>()
  var visitedNodes = 0

  fun visit(node: JsonElement?) {
    // X payloads can get very large, so cap the walk before defensive parsing turns expensive.
    if (node == null || visitedNodes >= MAX_FILTER_NODES || result.size >= MAX_FILTER_MATCHES) {
      return
    }

    visitedNodes += 1

    when (node) {
      is JsonArray -> {
        for (item in node) {
          visit(item)
          if (visitedNodes >= MAX_FILTER_NODES || result.size >= MAX_FILTER_MATCHES) {
            return
          }
        }
      }
      is JsonObject -> {
        if (node[key]?.stringValue() == value) {
          result += node
        }

        for (child in node.values) {
          visit(child)
          if (visitedNodes >= MAX_FILTER_NODES || result.size >= MAX_FILTER_MATCHES) {
            return
          }
        }
      }
      else -> Unit
    }
  }

  visit(data)
  return result
}

private fun getTimelineInstructions(payload: JsonElement): List<JsonElement> {
  val root = payload.jsonObjectOrNull()
  val candidates = listOfNotNull(
    root?.getObject("data")?.getObject("home")?.getObject("home_timeline_urt"),
    root?.getObject("home")?.getObject("home_timeline_urt"),
    root?.getObject("home_timeline_urt"),
    root,
  )

  for (candidate in candidates) {
    val instructions = candidate["instructions"].jsonArrayOrEmpty()
    if (instructions.isNotEmpty()) {
      return instructions
    }
  }

  return emptyList()
}

private fun getInstructionEntries(instruction: JsonObject): List<JsonObject> {
  val entries = instruction["entries"].jsonArrayOrEmpty().mapNotNull { it.jsonObjectOrNull() }.toMutableList()
  instruction["entry"].jsonObjectOrNull()?.let(entries::add)
  return entries
}

private fun extractBottomCursorFromNode(node: JsonElement?): String {
  val record = node.jsonObjectOrNull() ?: return ""
  val cursorType = record["cursorType"].stringValue()
  val value = record["value"].stringValue()
  if (cursorType == "Bottom" && value.isNotBlank()) {
    return value
  }

  val nestedContent = extractBottomCursorFromNode(record["content"])
  if (nestedContent.isNotBlank()) {
    return nestedContent
  }

  return extractBottomCursorFromNode(record["itemContent"])
}

private fun extractNextCursor(payload: JsonElement): String? {
  val instructions = getTimelineInstructions(payload)

  // The newest bottom cursor usually appears near the end of the instruction list.
  for (instructionIndex in instructions.indices.reversed()) {
    val instruction = instructions[instructionIndex].jsonObjectOrNull() ?: continue
    val entries = getInstructionEntries(instruction)
    for (entryIndex in entries.indices.reversed()) {
      val cursor = extractBottomCursorFromNode(entries[entryIndex])
      if (cursor.isNotBlank()) {
        return cursor
      }
    }
  }

  return findByFilter(payload, "cursorType", "Bottom").firstOrNull()?.get("value")?.stringValue()?.takeIf { it.isNotBlank() }
}

private fun extractTweet(timelineTweetNode: JsonObject): XTimelineItem? {
  val tweetResults = timelineTweetNode.getObject("tweet_results")
  // X wraps tweets in a few result shapes, so normalize that before reading legacy fields.
  val result = unwrapTweetResult(tweetResults?.getObject("result")) ?: return null
  val legacy = result.getObject("legacy") ?: return null
  val id = result["rest_id"].stringValue()
  if (id.isBlank()) {
    return null
  }

  val noteTweetResult = result.getObject("note_tweet")
    ?.getObject("note_tweet_results")
    ?.getObject("result")

  val core = result.getObject("core")
  val userResult = core?.getObject("user_results")?.getObject("result")
  val userLegacy = userResult?.getObject("legacy")
  val userCore = userResult?.getObject("core")
  val authorHandle = userCore?.get("screen_name").stringValue().ifBlank {
    userLegacy?.get("screen_name").stringValue()
  }

  val quotedResult = unwrapTweetResult(
    result.getObject("quoted_status_result")?.getObject("result"),
  )

  val retweetedResult = legacy.getObject("retweeted_status_result")?.getObject("result")
  val retweetedTweet = retweetedResult?.getObject("tweet")
  val isRetweet = retweetedResult?.get("rest_id").stringValue().isNotBlank() ||
    retweetedTweet?.get("rest_id").stringValue().isNotBlank()
  val isQuote = legacy["is_quote_status"].booleanValue() || quotedResult?.get("rest_id").stringValue().isNotBlank()

  val quotedLegacy = quotedResult?.getObject("legacy")
  val quotedNoteResult = quotedResult?.getObject("note_tweet")
    ?.getObject("note_tweet_results")
    ?.getObject("result")
  val quotedUserResult = quotedResult?.getObject("core")
    ?.getObject("user_results")
    ?.getObject("result")
  val quotedUserLegacy = quotedUserResult?.getObject("legacy")
  val quotedUserCore = quotedUserResult?.getObject("core")

  val quotedAuthorHandle = quotedUserCore?.get("screen_name").stringValue().ifBlank {
    quotedUserLegacy?.get("screen_name").stringValue()
  }

  return XTimelineItem(
    id = id,
    text = noteTweetResult?.get("text").stringValue().ifBlank { legacy["full_text"].stringValue() },
    createdAt = toIsoDate(legacy["created_at"].stringValue()),
    authorName = userCore?.get("name").stringValue().ifBlank { userLegacy?.get("name").stringValue() },
    authorHandle = authorHandle,
    lang = legacy["lang"].stringValue(),
    replyTo = legacy["in_reply_to_status_id_str"].stringValue(),
    quoteCount = legacy["quote_count"].intValue(),
    replyCount = legacy["reply_count"].intValue(),
    retweetCount = legacy["retweet_count"].intValue(),
    likeCount = legacy["favorite_count"].intValue(),
    viewCount = result.getObject("views")?.get("count")?.intValueOrNull(),
    isRetweet = isRetweet,
    isQuote = isQuote,
    quotedText = quotedNoteResult?.get("text").stringValue().ifBlank { quotedLegacy?.get("full_text").stringValue() },
    quotedLang = quotedLegacy?.get("lang").stringValue(),
    quotedAuthorHandle = quotedAuthorHandle,
    quotedMedia = extractMedia(quotedLegacy),
    url = if (authorHandle.isBlank()) "" else "https://x.com/$authorHandle/status/$id",
    media = extractMedia(legacy),
  )
}

private fun extractMedia(legacy: JsonObject?): List<XTimelineMedia> {
  val mediaEntries = legacy?.getObject("extended_entities")?.get("media").jsonArrayOrEmpty()

  return mediaEntries.mapNotNull { entry ->
    val media = entry.jsonObjectOrNull() ?: return@mapNotNull null
    val type = mediaTypeFromRaw(media["type"].stringValue())
    val mediaUrl = extractBestMediaUrl(media)
    val expandedUrl = media["expanded_url"].stringValue()

    if (mediaUrl.first.isBlank() && expandedUrl.isBlank()) {
      return@mapNotNull null
    }

    XTimelineMedia(
      type = type,
      url = mediaUrl.first,
      expandedUrl = expandedUrl,
      thumbnailUrl = mediaUrl.second,
    )
  }
}

private fun extractBestMediaUrl(media: JsonObject): Pair<String, String?> {
  return when (mediaTypeFromRaw(media["type"].stringValue())) {
    XTimelineMediaType.PHOTO -> media["media_url_https"].stringValue() to null
    else -> {
      val variants = media.getObject("video_info")?.get("variants").jsonArrayOrEmpty()
      var selectedUrl = ""
      var selectedBitrate = -1

      for (variantElement in variants) {
        val variant = variantElement.jsonObjectOrNull() ?: continue
        val contentType = variant["content_type"].stringValue()
        val url = variant["url"].stringValue()
        val bitrate = variant["bitrate"].intValue()

        if (url.isBlank()) {
          continue
        }

        if (contentType.contains("mp4") && bitrate >= selectedBitrate) {
          selectedUrl = url
          selectedBitrate = bitrate
          continue
        }

        if (selectedUrl.isBlank()) {
          selectedUrl = url
        }
      }

      selectedUrl to media["media_url_https"].stringValue().ifBlank { null }
    }
  }
}

private fun mediaTypeFromRaw(value: String): XTimelineMediaType {
  return when (value) {
    "photo" -> XTimelineMediaType.PHOTO
    "video" -> XTimelineMediaType.VIDEO
    "animated_gif" -> XTimelineMediaType.ANIMATED_GIF
    else -> XTimelineMediaType.UNKNOWN
  }
}

private fun unwrapTweetResult(value: JsonObject?): JsonObject? {
  if (value == null) {
    return null
  }

  return if (value["__typename"].stringValue() == "TweetWithVisibilityResults") {
    value.getObject("tweet")
  } else {
    value
  }
}

private fun toIsoDate(rawDate: String): String {
  if (rawDate.isBlank()) {
    return ""
  }

  return try {
    OffsetDateTime.parse(rawDate, xDateFormatter).toInstant().toString()
  } catch (_: Exception) {
    try {
      Instant.parse(rawDate).toString()
    } catch (_: Exception) {
      ""
    }
  }
}

private fun JsonElement?.jsonObjectOrNull(): JsonObject? = this as? JsonObject

private fun JsonElement?.jsonArrayOrEmpty(): JsonArray = this as? JsonArray ?: JsonArray(emptyList())

private fun JsonObject.getObject(key: String): JsonObject? = this[key].jsonObjectOrNull()

private fun JsonElement?.stringValue(): String {
  return (this as? JsonPrimitive)?.contentOrNull.orEmpty()
}

private fun JsonElement?.intValue(): Int {
  return (this as? JsonPrimitive)?.intOrNull ?: stringValue().toIntOrNull() ?: 0
}

private fun JsonElement?.intValueOrNull(): Int? {
  return (this as? JsonPrimitive)?.intOrNull ?: stringValue().toIntOrNull()
}

private fun JsonElement?.booleanValue(): Boolean {
  return (this as? JsonPrimitive)?.booleanOrNull ?: false
}
